src/main/java/com/endava/internship/collections/StudentList.java

```diff
@@ -1,145 +1,332 @@
+            }
+            return StudentList.this.get(currentElement++);
+        }
+
+        @Override
+        public void remove() {
+            StudentList.this.remove(elements[--currentElement]);
+        }
     }
 
     @Override
     public Object[] toArray() {
-        //TODO
-        return null;
+        return Arrays.copyOf(elements, size);
     }
 
+    @SuppressWarnings("unchecked")
     @Override
-    public <T> T[] toArray(T[] ts) {
-        //TODO
-        return null;
+    public <E> E[] toArray(E[] ts) {
+        if (Objects.isNull(ts)) {
+            throw new NullPointerException();
+        }
+        if (ts.length >= this.size()) {
+            int i = 0;
+            for (; i < this.size; i++) {
+                ts[i] = (E) this.get(i);
+            }
+            for (; i < ts.length; i++) {
+                ts[i] = null;
+            }
+            return ts;
+        }
+
+        return (E[]) Arrays.copyOf(elements, size, ts.getClass());
     }
 
     @Override
-    public boolean add(Student student) {
-        //TODO
-        return false;
+    public boolean add(T t) {
+        if (size == elements.length) {
+            resizeArray();
+        }
+        elements[size++] = t;
+        return true;
     }
 
     @Override
-    public boolean remove(Object o) {
-        //TODO
+    public boolean remove(Object t) {
+        if(!contains(t)){
+          return false;
+        }
+        int index = this.indexOf(t);
+        if (index >= 0) {
+            remove(index);
+            return true;
+        }
         return false;
     }
 
     @Override
     public void clear() {
-        //TODO
+        for (int i = 0; i < size; i++) {
+            elements[i] = null;
+        }
+        size = 0;
     }
 
+    @SuppressWarnings("unchecked")
     @Override
-    public Student get(int i) {
-        //TODO
-        return null;
+    public T get(int i) {
+        if (i >= 0 && i < size) {
+            return (T) elements[i];
+        } else {
+            throw new IndexOutOfBoundsException("Insertion index was out of range. Must be non-negative and less than or equal to size");
+        }
     }
 
+    @SuppressWarnings("unchecked")
     @Override
-    public Student set(int i, Student student) {
-        //TODO
-        return null;
+    public T set(int i, T t) {
+        if (i >= 0 && i < size) {
+            return (T) (elements[i] = t);
+        } else {
+            throw new IndexOutOfBoundsException("Insertion index was out of range. Must be non-negative and less than or equal to size");
+        }
     }
 
     @Override
-    public void add(int i, Student student) {
-        //TODO
+    public void add(int i, T t) {
+        if (i < 0 || i > size) {
+            throw new IndexOutOfBoundsException("Insertion index was out of range. Must be non-negative and less than or equal to size");
+        }
+        if (size == elements.length) {
+            resizeArray();
+        }
+        if (size - i >= 0) {
```

Comment:
how many elements do you copy when   `size - i == 0 `?

Reply:
One

Reply:
Done.